### PR TITLE
very basic token creation

### PR DESCRIPTION
**Pull request by gr2m**

### CODE REVIEW

File: lib/routes/requests.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
+var Boom = require('boom')
+
+/* NOTES:
+  * What Auth do we use against CouchDB to get/set user doc? (user not logged in)
```
Comment (gr2m) [inline lib/routes/requests.js]: You can copy how it's done in the `GET /session/account` handler here: https://github.com/hoodiehq/hapi-couchdb-account-api/blob/master/lib/routes/account.js#L28-L33

We do a straight request against couchDB, and set the cookie from the bearer token passed in the `Authorization` header. Note that the API does not expose the entire user doc, only id (not `_id` but a custom generated id that always remains the same, even if a username changes) and username. So you can safely add a `token` property to the _usersdoc.

File: lib/routes/requests.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  * What url do we send to the user? (how do configure url)
```
Comment (gr2m) [inline lib/routes/requests.js]: The origin should be in `server.info.uri`. See also the `notifications.templates.passwordreset` option. I think that should work, right?

File: lib/routes/requests.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  * Could we use promises? I think it will clean the code up nicely
```
Comment (gr2m) [inline lib/routes/requests.js]: We usually stick to callbacks in node code as it feels more natural for most node developers, but this is work in progress code, and especially as this is your first contribution, I'd say use whatever you are most comfortable with, we can still change things later once it's working, no big deal at all :)

File: tests/unit/routes/requests-test.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        .reply(200, {
+          uuids: ['9aef3ffae9597ce60dfec0fee200094c']
+        })
+        .put('/_users/org.couchdb.user%3Ajohn', function (body) {
+          return body.reset_token === '9aef3ffae9597ce60dfec0fee200094c'
+        })
+        .reply(200)
+    }
+
+    group.test('Creates token and timestamp for username', function (t) {
```
Comment (gr2m) [inline tests/unit/routes/requests-test.js]: Very good point with the timestamp. For handling timeouts, I guess the simplest thing would be to set timers in the request handlers, plus an initial bootstrap on startup?

File: lib/routes/requests.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  function createPasswordReset (doc, cb) {
+    request.get({
+      url: '/uuids'
+    }, function (err, resp, body) {
+      if (err) {
+        return cb(err)
+      }
+      doc.reset_time_stamp = new Date()
+      doc.reset_token = body.uuids[0]
```
Comment (gr2m) [inline lib/routes/requests.js]: just a minor (feel free to ignore we can fix this later): we use camelCase for variables, properties and methods

File: lib/routes/requests.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    })
+  }
+
+  server.route([{
+    method: 'POST',
+    path: prefix + '/requests',
+    handler: function (req, reply) {
+      var username = req.payload.username
+      // TODO: this fails due to
+      // https://github.com/pouchdb/express-pouchdb/issues/240
```
Comment (gr2m) [inline lib/routes/requests.js]: This should be resolved now, I'll check tomorrow

**Comment by gr2m** [comment]
@garrensmith that's great work so far!

The idea is, that an email then gets sent out with a URL pointing to my app, which passes the token, so I can reset my password with a `PATCH /session/account`, and a `Authorization: Token uuid123resetTokenHere` Header. And if that succeeds, I can sign in with the new password

**Comment by gr2m** [comment]
@garrensmith we started brainstorming on a admin client API that could be used within the server as well (in the route handlers), would be great to get your feedback on it:

- Pull Request: https://github.com/hoodiehq/account-client/pull/4
- WIP Version: https://github.com/hoodiehq/account-client/blob/admin-client/admin/README.md

We have an open PR that implements the new `/session` and `/session/account` routes which uses that API already: https://github.com/hoodiehq/hapi-couchdb-account-api/blob/json-api/routes/account.js#L37-L49

For the password reset workflow, we'd use the `POST /requests` route twice

First request would be a `passwordreset`, and the handler (passed as `options.requests.passwordreset` to the hapi plugin)

```js
function passwordReset(request, reply) {
  var server = request.connection.server
  var api = server.plugins.account.api
  var email = request.payload.data.attributes.email
  var promise = api.requests.add({type: 'passwordreset', contact: contact}, options)
  .then(sendEmailWithToken)
  reply(promise)
}
```

Then the user would click the password reset link and would submit a form with the new password, which would send another request to `POST /requests`, this time with type `accountupdate`, the handler would look like this

```js
function updateAccount(request, reply) {
  var server = request.connection.server
  var api = server.plugins.account.api
  var token = request.payload.data.attributes.token
  var password = request.payload.data.attributes.password
  var promise = api.accounts.update({
    password: password
  }, {
    auth: {
      token: token
    }
  )
  reply(promise)
}
```

The `api.accounts.update` request would be authenticated with the one-time token (tokens are stored in the `_users` doc. It would update the account and invalidate the token.

So far the theory. What do you think? Does that all make sense?

**Comment by garrensmith** [comment]
@gr2m that looks really cool. I left a small comment on the one.

When is a good time to integrate the password reset code into the new json-api code?

**Comment by gr2m** [comment]
We have a big PR for the JSON API here: Patricia will review it soon, but feel free to have a look, too: https://github.com/hoodiehq/hapi-couchdb-account-api/pull/6. It adds handlers for the session & account routes. Once that is in, putting in your password reset could be the next step

**Comment by gr2m** [comment]
I'm sorry I closed the PR & deleted the branch by accident. I had to throw away most of the code from master as part of https://github.com/hoodiehq/hoodie/issues/436, but if you push your branch again, I can incorporate your work so we don't have to start from scratch?

**Comment by garrensmith** [comment]
@gr2m no problem the branch is still here https://github.com/garrensmith/hapi-couchdb-account-api/tree/reset-password

**Comment by gr2m** [comment]
Sorry I didn't get back on this yet, it’s not forgotten. The opposite, I'm excited to put it back in as soon as we have the `/requests` API implemented. Right now I'm busy with the `/accounts` API for admins, but doing some good progress there:
- https://github.com/hoodiehq/hoodie-server-account/issues/12
- https://github.com/hoodiehq/hoodie-client-account/issues/7

I work on these because the account module will now be also used in https://github.com/ehealthafrica/kazana, which brings it to all kind of interesting places, I hope to share more on this soon :)

Anyway, just a quick update on this, I want to make sure you don’t feel like your work wouldn’t make it into the account module, it just takes a bit more time :)

**Comment by garrensmith** [comment]
no problem @gr2m I'm pretty busy with pouchdb-find that I wouldn't have any time to help so. I don't mind at all. Awesome work on the admin system. That is really exciting.

**Pull request closed without merging**